Feature: state:modified.macros

jtcohen6 opened the pull request:
resolves #2704
resolves #3278
resolves [this tweet](https://twitter.com/tjwaterman99/status/1420809652548296704)

### Description

If a macro's SQL is modified, mark any node that depends on it (or depends on a macro that depends on it) as modified.

While I was here, it was trivial to make sub-selectors (finally) happen:
- `state:modified.body`
- `state:modified.configs`
- `state:modified.persisted_descriptions`
- ~`state:modified.database_representations`~ `state:modified.relation`
- `state:modified.macros`

### For review
- I don't really know how to write recursive functions :)
- Let's revisit some of the nomenclature here (vs. the original proposal in #2704). Even though it's a mouthful, I prefer `persisted_descriptions` vs. just `descriptions`, to be clear it's the subset with `persist_docs` enabled. On the other hand, I'm not sure about `modified.database_representations` vs. something shorter & sweeter like `modified.relations`.
- There are a few more `same_x` checks (`same_fqn`, `same_quoting`, `same_external`) that are resource-specific and wouldn't have dedicated subselectors in this first round. I think that's ok—these aren't ones we've heard as much demand for, and they're still included in the `state:modified` superset—just worth calling out.

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

nathaniel-may commented on core/dbt/graph/selector_methods.py:
```diff
@@ -394,44 +392,71 @@ def _macros_modified(self) -> List[str]:
+    def check_modified_macros(self, old: Optional[SelectorTarget], new: SelectorTarget) -> bool:
+        return self.recursively_check_macros_modified(new)
+
+    def check_new(self, old: Optional[SelectorTarget], new: SelectorTarget) -> bool:
         return old is None
```
I know this was here already, but it seems pretty unnecessary. Maybe change `L471` to the function body of this method.

nathaniel-may commented on core/dbt/graph/selector_methods.py:
```diff
@@ -443,8 +468,13 @@ def search(
 
         state_checks = {
-            'modified': self.check_modified,
             'new': self.check_new,
```
```suggestion
            # maybe with some sort of comment here.
            'new': old is None,
```

kwigley replied:
I think this needs to be a function/lambda still. maybe?
```suggestion
            'new': lambda old, _: old is None,
```

nathaniel-may replied:
ah yeah that might be right.

jtcohen6 replied:
lambda's a neat idea. I'll leave a comment, too

nathaniel-may commented on test/integration/062_defer_state_test/changed_models/table_model.sql:
```diff
@@ -1,2 +1,3 @@
 {{ config(materialized='table') }}
 select * from {{ ref('ephemeral_model') }}
+-- {{ my_macro() }}
```
Just want to double check that these should be sql commented but not jinja commented.

jtcohen6 replied:
Yup! I still want the macro to be picked up as a dependency, just without actually doing anything. I can make this clearer with a comment

kwigley commented on core/dbt/graph/selector_methods.py:
```diff
@@ -394,44 +392,71 @@ def _macros_modified(self) -> List[str]:
+        else:
+            return False
+
+    def check_modified_macros(self, old: Optional[SelectorTarget], new: SelectorTarget) -> bool:
+        return self.recursively_check_macros_modified(new)
```
Unfortunately, this function is called with 2 args by default
https://github.com/dbt-labs/dbt/blob/0efaaf7dafccb3329ade084a14218e73e8e83b08/core/dbt/graph/selector_methods.py#L468-L469
^^ `checker` here is assigned a function by looking up in this map
https://github.com/dbt-labs/dbt/blob/4fe16a67e89ee0d5ba1602358d97fa60e9634b87/core/dbt/graph/selector_methods.py#L470-L478
Dunno the best way to handle this, but a pythonic convention for args or variables that won't be used is `_` 🤷 

```suggestion
    def check_modified_macros(self, _: Optional[SelectorTarget], new: SelectorTarget) -> bool:
        return self.recursively_check_macros_modified(new)
```

nathaniel-may replied:
yeah I'll take the underscore given the situation. But also, oof.

kwigley replied:
oof is right

jtcohen6 replied:
pythonic it is!

kwigley reviewed (approved):
nice! looks good

jtcohen6 commented:
Thanks for the reviews @nathaniel-may @kwigley! I added some comments where the code wasn't clear.

I ended up renaming `state:modified.database_representations` to `state:modified.relation`. I think I like it better that way. Relation is a wonkier word, but it's [one that dbt users should be familiar with](https://docs.getdbt.com/reference/dbt-classes#relation) as indicating `database.schema.table`.

I spent slightly too long thinking about which of these sub-selector names should be singular and which should be plural. I decided to use "per node" as the frame, such that:
- one `body` per node
- multiple `configs` per node
- multiple `persisted_descriptions` per node (potentially relation + columns)
- one `relation` per node
- multiple `macros` that a node may depend on
